### PR TITLE
store: add error code for assertion

### PR DESCRIPTION
**Pull request by ekexium**
### What problem does this PR solve?

Issue Number: ref #26833 

Problem Summary:

- Add an error code for assertion.
- Fetch the MVCC history for assertion failure.

### What is changed and how it works?

The error log would look like
```log
[2022/01/17 20:46:37.202 +08:00] [ERROR] [session.go:616] ["assertion failed"] [message="[tikv:8141]assertion failed: key: 74800000000000003c5f72013300000000000000f8, assertion: start_ts:430551491464921091 key:\"t\\200\\000\\000\\000\\000\\000\\000<_r\\0013\\000\\000\\000\\000\\000\\000\\000\\370\" assertion:NotExist existing_start_ts:430551491464658947 existing_commit_ts:430551491464658948 , start_ts: 430551491464921091, existing start ts: 430551491464658947, existing commit ts: 430551491464658948"] ["mvcc history"="{\"decoded\":{\"430551491463610371\":{\"v\":\"30\",\"v2\":\"300\",\"v3\":\"3000\",\"v4\":\"30000\"},\"430551491463872515\":{\"v\":\"30\",\"v2\":\"300\",\"v3\":\"3000\",\"v4\":\"30000\"},\"430551491464134659\":{\"v\":\"30\",\"v2\":\"300\",\"v3\":\"3000\",\"v4\":\"30000\"},\"430551491464658947\":{\"v\":\"30\",\"v2\":\"300\",\"v3\":\"3000\",\"v4\":\"30000\"}},\"key\":\"74800000000000003C5F72013300000000000000F8\",\"mvcc\":{\"info\":{\"writes\":[{\"start_ts\":430551491464658947,\"commit_ts\":430551491464658948,\"short_value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"},{\"type\":1,\"start_ts\":430551491464658945,\"commit_ts\":430551491464658946},{\"type\":3,\"start_ts\":430551491464396804,\"commit_ts\":430551491464396804},{\"start_ts\":430551491464134659,\"commit_ts\":430551491464134660,\"short_value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"},{\"type\":1,\"start_ts\":430551491463872517,\"commit_ts\":430551491464134656},{\"start_ts\":430551491463872515,\"commit_ts\":430551491463872516,\"short_value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"},{\"type\":1,\"start_ts\":430551491463610373,\"commit_ts\":430551491463872512},{\"start_ts\":430551491463610371,\"commit_ts\":430551491463610372,\"short_value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"}],\"values\":[{\"start_ts\":430551491464658947,\"value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"},{\"start_ts\":430551491464658945},{\"start_ts\":430551491464396804},{\"start_ts\":430551491464134659,\"value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"},{\"start_ts\":430551491463872517},{\"start_ts\":430551491463872515,\"value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"},{\"start_ts\":430551491463610373},{\"start_ts\":430551491463610371,\"value\":\"gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA==\"}]}},\"regionID\":2}"] [stack="github.com/pingcap/tidb/session.(*session).handleAssertionFailure\n\t/Users/eke/code/tidb/session/session.go:616\ngithub.com/pingcap/tidb/session.(*session).doCommit\n\t/Users/eke/code/tidb/session/session.go:568\ngithub.com/pingcap/tidb/session.(*session).doCommitWithRetry\n\t/Users/eke/code/tidb/session/session.go:758\ngithub.com/pingcap/tidb/session.(*session).CommitTxn\n\t/Users/eke/code/tidb/session/session.go:820\ngithub.com/pingcap/tidb/session.autoCommitAfterStmt\n\t/Users/eke/code/tidb/session/tidb.go:272\ngithub.com/pingcap/tidb/session.finishStmt\n\t/Users/eke/code/tidb/session/tidb.go:238\ngithub.com/pingcap/tidb/session.runStmt\n\t/Users/eke/code/tidb/session/session.go:1766\ngithub.com/pingcap/tidb/session.(*session).ExecuteStmt\n\t/Users/eke/code/tidb/session/session.go:1639\ngithub.com/pingcap/tidb/testkit.(*TestKit).Exec\n\t/Users/eke/code/tidb/testkit/testkit.go:145\ngithub.com/pingcap/tidb/testkit.(*TestKit).ExecToErr\n\t/Users/eke/code/tidb/testkit/testkit.go:181\ngithub.com/pingcap/tidb/table/tables_test.TestTxnAssertion.func1\n\t/Users/eke/code/tidb/table/tables/tables_test.go:781\ngithub.com/pingcap/tidb/table/tables_test.TestTxnAssertion.func4.6\n\t/Users/eke/code/tidb/table/tables/tables_test.go:890\ngithub.com/pingcap/tidb/table/tables_test.TestTxnAssertion.func2\n\t/Users/eke/code/tidb/table/tables/tables_test.go:789\ngithub.com/pingcap/tidb/table/tables_test.TestTxnAssertion.func4\n\t/Users/eke/code/tidb/table/tables/tables_test.go:889\ngithub.com/pingcap/tidb/table/tables_test.TestTxnAssertion\n\t/Users/eke/code/tidb/table/tables/tables_test.go:916\ntesting.tRunner\n\t/opt/homebrew/Cellar/go/1.17.5/libexec/src/testing/testing.go:1259"]
```

which contains the mvcc history:
```json
{
    "decoded": {
        "430551491463610371": {
            "v": "30",
            "v2": "300",
            "v3": "3000",
            "v4": "30000"
        },
        "430551491463872515": {
            "v": "30",
            "v2": "300",
            "v3": "3000",
            "v4": "30000"
        },
        "430551491464134659": {
            "v": "30",
            "v2": "300",
            "v3": "3000",
            "v4": "30000"
        },
        "430551491464658947": {
            "v": "30",
            "v2": "300",
            "v3": "3000",
            "v4": "30000"
        }
    },
    "key": "74800000000000003C5F72013300000000000000F8",
    "mvcc": {
        "info": {
            "writes": [
                {
                    "start_ts": 430551491464658947,
                    "commit_ts": 430551491464658948,
                    "short_value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                },
                {
                    "type": 1,
                    "start_ts": 430551491464658945,
                    "commit_ts": 430551491464658946
                },
                {
                    "type": 3,
                    "start_ts": 430551491464396804,
                    "commit_ts": 430551491464396804
                },
                {
                    "start_ts": 430551491464134659,
                    "commit_ts": 430551491464134660,
                    "short_value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                },
                {
                    "type": 1,
                    "start_ts": 430551491463872517,
                    "commit_ts": 430551491464134656
                },
                {
                    "start_ts": 430551491463872515,
                    "commit_ts": 430551491463872516,
                    "short_value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                },
                {
                    "type": 1,
                    "start_ts": 430551491463610373,
                    "commit_ts": 430551491463872512
                },
                {
                    "start_ts": 430551491463610371,
                    "commit_ts": 430551491463610372,
                    "short_value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                }
            ],
            "values": [
                {
                    "start_ts": 430551491464658947,
                    "value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                },
                {
                    "start_ts": 430551491464658945
                },
                {
                    "start_ts": 430551491464396804
                },
                {
                    "start_ts": 430551491464134659,
                    "value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                },
                {
                    "start_ts": 430551491463872517
                },
                {
                    "start_ts": 430551491463872515,
                    "value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                },
                {
                    "start_ts": 430551491463610373
                },
                {
                    "start_ts": 430551491463610371,
                    "value": "gAAEAAAAAgMEBQEAAwAFAAoAHiwBuAszMDAwMA=="
                }
            ]
        }
    },
    "regionID": 2
}
```

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by ekexium** [comment]
PTAL @MyonKeminta @cfzjywxk

**Review by MyonKeminta — COMMENTED** [review]
Ah, thanks.
Should we also trigger mvcc information collecting in the reporter?

**Comment by ekexium** [comment]
> Ah, thanks. Should we also trigger mvcc information collecting in the reporter?

Maybe it would be better. It seems not very convenient to get store-related information from where the error occurred?
If the mvcc collection is always triggered, how about returning it in the assertion message?

**Comment by cfzjywxk** [comment]
> > Ah, thanks. Should we also trigger mvcc information collecting in the reporter?
> 
> Maybe it would be better. It seems not very convenient to get store-related information from where the error occurred? If the mvcc collection is always triggered, how about returning it in the assertion message?

The new inconsistency reporter seems to have such abilities to get the mvcc versions.

**Comment by ekexium** [comment]
> The new inconsistency reporter seems to have such abilities to get the mvcc versions.

Yes, I mean the reporter(or the helper it uses) requires a `storage` while the storage information is invisible in the txn driver.

File: session/session.go [hunk 1]
```diff
@@ -677,6 +684,59 @@ func (c *cachedTableRenewLease) commitTSCheck(commitTS uint64) bool {
...
 
+// handleAssertionFailure extracts the possible underlying assertionFailed error,
+// gets the corresponding MVCC history and logs it.
+// If it's not an assertion failure, returns the original error.
+func (s *session) handleAssertionFailure(err error) error {
+	var assertionFailure *tikverr.ErrAssertionFailed
+	if stderrs.As(err, &assertionFailure) {
+		key := assertionFailure.Key
+		newErr := kv.ErrAssertionFailed.GenWithStackByArgs(
+			hex.EncodeToString(key), assertionFailure.String(), assertionFailure.StartTs,
```
[the remark this answers is not in the document]
Comment (ekexium) [reply session/session.go:695]: Ah yes

File: session/session.go [hunk 2]
```diff
@@ -689,52 +691,55 @@ func (c *cachedTableRenewLease) commitTSCheck(commitTS uint64) bool {
...
+		return err
+	}
+	key := assertionFailure.Key
+	newErr := kv.ErrAssertionFailed.GenWithStackByArgs(
+		hex.EncodeToString(key), assertionFailure.Assertion.String(), assertionFailure.StartTs,
+		assertionFailure.ExistingStartTs, assertionFailure.ExistingCommitTs,
+	)
+
+	if s.GetSessionVars().EnableRedactLog {
+		return newErr
```
Comment (MyonKeminta) [inline session/session.go:704]: Hope the transaction event table can be supported someday so we have place to save the details :(

File: session/session.go [hunk 3]
```diff
@@ -689,52 +691,55 @@ func (c *cachedTableRenewLease) commitTSCheck(commitTS uint64) bool {
...
+	}
+
+	var decodeFunc func(kv.Key, *kvrpcpb.MvccGetByKeyResponse, map[string]interface{})
+	// if it's a record key or an index key, decode it
+	if infoSchema, ok := s.sessionVars.TxnCtx.InfoSchema.(infoschema.InfoSchema); ok &&
+		infoSchema != nil && (tablecodec.IsIndexKey(key) || tablecodec.IsIndexKey(key)) {
```
Comment (MyonKeminta) [inline session/session.go:710]: Did you mean `(tablecodec.IsIndexKey(key) || tablecodec.IsRecordKey(key))` ? And why it needs to be added here?

Comment (ekexium) [reply session/session.go]: 😓..my bad. To avoid the duplication of the code that finds the table and handle errors.

**Review by MyonKeminta — APPROVED** [review]

File: session/session.go [hunk 4]
```diff
@@ -677,6 +686,62 @@ func (c *cachedTableRenewLease) commitTSCheck(commitTS uint64) bool {
...
+				}
+				decodeFunc = consistency.DecodeIndexMvccData(indexInfo)
+			}
+		} else {
+			logutil.BgLogger().Warn("assertion failed but table not found in infoschema", zap.Int64("tableID", tableID))
+		}
+	}
+	if store, ok := s.store.(helper.Storage); ok {
+		content := consistency.GetMvccByKey(store, key, decodeFunc)
+		logutil.BgLogger().Error("assertion failed", zap.String("message", newErr.Error()), zap.String("mvcc history", content))
```
Comment (cfzjywxk) [inline session/session.go:740]: We could pass down the `ctx` parameter so the error log contains connection id.

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**